<a href="https://colab.research.google.com/github/rayoen0/data-analytics-with-python-classroom-1a8e1c-individual-assignment-iii-Case-Assignment-III/blob/main/Assignment_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

IMPORTANT: Before you start, enter your name and student number below.

**Full Name**: Rayane El Metni

**Student Number**: 400677235

# Predictive Analytics for Nata Supermarket

Welcome to Part III of our case assignment. In this part, we will continue working with the same dataset of **Nata Supermarket**.  Our focus here will be on performing predictive modeling tasks.

Throughout this assignment, please ensure that your results are reproducible by setting the **random_state to 20**



# Loading and preparing the data

To begin with, load the data as a `pandas` data frame. Recall that you there are missing values in the data. **Make sure to address** the following issues from part I of the assignment before starting your analysis:

* Remove the missing values
* Remove any column of constant values
* Convert the column `Dt_Customer` to number of days the customer has been with the company.

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, KFold, cross_val_score, GridSearchCV
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.metrics import (mean_squared_error, r2_score, mean_absolute_error,
                             classification_report, confusion_matrix)
import plotly.express as px

In [18]:
from datetime import datetime

# Load the dataset
df = pd.read_csv("/content/W33836-XLS-ENG.csv", sep=";")

# Removing missing values
df.dropna(inplace=True)

# Checking for constant columns and removing them as these don't provide predictive value
constant_cols = df.columns[df.nunique() <= 1]
print(f"Constant columns: {constant_cols}")
df = df.drop(columns=constant_cols) # Removed inplace=True and reassigned df


# 3. Convert Dt_Customer to number of days with company
# First, convert to datetime
df['Dt_Customer'] = pd.to_datetime(df['Dt_Customer'], format='mixed', dayfirst=False)
# Use a reference date, we'll use today to compare
reference_date = datetime.today()
df['Dt_Customer'] = (reference_date - df['Dt_Customer']).dt.days

df.head()

Constant columns: Index(['Z_CostContact', 'Z_Revenue'], dtype='object')


ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

   Dt_Customer  Recency  MntWines  ...  NumCatalogPurchases  \
0         4853       58       635  ...                   10   
1         4303       38        11  ...                    1   
2         4502       26       426  ...                    2   
3         4329       26        11  ...                    0   
4         4351       94       173  ...                    3   

   NumStorePurchases  NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  \
0                  4                  7             0             0   
1                  2                  5             0             0   
2                 10                  4             0             0   
3                  4                  6             0             0   
4                  6                  5             0             0   

   AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  Response  
0             0             0             0         0         1  
1             0             0             0         0         0  
2             0             0             0         0         0  
3             0             0             0         0         0  
4             0             0             0         0         0  

[5 rows x 27 columns]

# Question 1 (45 points)

In this question, we will predict a customer's spending amount on each product category over a two year period. Let us assume that when we try to predict a customer's spending on a product category (such as wines), their spending on other products is not observable.

In this question and Question 2, we will focus on **wines**.

(i). **Split the data** into two data frames, X (**features**) and y (**target**).

Then, further **split the data** into **training** and **testing** sets.

In [20]:
# Define target variable. we're predicting wine spending
y = df['MntWines']

# Define features - exclude all spending columns and ID
# When predicting wines, we assume other product spending is not observable
spending_cols = ['MntWines', 'MntFruits', 'MntMeatProducts',
                 'MntFishProducts', 'MntSweetProducts', 'MntGoldProds']
X = df.drop(columns=spending_cols + ["ID"])  # Remove ID as well since it's not predictive

print(f"Feature columns: {list(X.columns)}")
print(f"X shape: {X.shape}, y shape: {y.shape}")

# Split into training and testing sets (80-20 split)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=20
)

print(f"Training set size: {X_train.shape[0]}")
print(f"Testing set size: {X_test.shape[0]}")


Feature columns: ['Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 'Recency', 'NumDealsPurchases', 'NumWebPurchases', 'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth', 'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1', 'AcceptedCmp2', 'Complain', 'Response']
X shape: (2216, 20), y shape: (2216,)
Training set size: 1772
Testing set size: 444


(ii). **Build a machine learning pipeline** that combines the following steps to predict spending amount on wines:

* Performing one-hot encoding for the categorical features;
* A random forest model for regression.

In the random forest model, **specify** the following hyperparameters:
* Number of trees;
* Maximum depth of any tree
* Minimum number of data points required to split a node;
* Minimum number of data points in any leaf node

In addition, **fit your model** to the training data.

In [23]:
# First, we identify categorical and numerical columns
cat_cols = ['Education', 'Marital_Status']  # Categorical features
num_cols = ['Year_Birth', 'Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 'Recency', 'NumDealsPurchases', 'NumWebPurchases', 'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth', 'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1', 'AcceptedCmp2', 'Complain', 'Response']  # The rest are numerical features
#num_cols = [col for col in X.columns if col not in cat_cols]

print(f"Categorical features: {cat_cols}")
print(f"Numerical features: {num_cols}")

# Create the column transformer for preprocessing
# We'll use one-hot encoding for categorical variables
# For numerical variables, we'll pass them through without scaling since
# tree-based models like Random Forest are robust and not sensitive to outliers
ct = ColumnTransformer(
    transformers=[
        ("cat", OneHotEncoder(handle_unknown='ignore', sparse_output=False, drop="first"), cat_cols)
    ],
    remainder="passthrough"  # Pass through numerical columns unchanged
)

# Define the Random Forest model with specified hyperparameters
# Based on the lecture materials, we specify:
# - Number of trees: Using 100 as a reasonable starting point
# - Maximum depth: 10 to prevent overfitting
# - Min samples to split: 10 to ensure nodes have enough data
# - Min samples per leaf: 5 for stable predictions
rf = RandomForestRegressor(
    n_estimators=100,      # Number of trees in the forest
    max_depth=10,          # Maximum depth of any tree
    min_samples_split=10,  # Minimum data points required to split a node
    min_samples_leaf=5,    # Minimum data points in any leaf node
    random_state=20
)

# Create the pipeline combining preprocessing and model
pipe = Pipeline(steps=[
    ("preprocess", ct),
    ("model", rf)
])

# Fit the model to training data
pipe.fit(X_train, y_train)

pipe

Categorical features: ['Education', 'Marital_Status']
Numerical features: ['Year_Birth', 'Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 'Recency', 'NumDealsPurchases', 'NumWebPurchases', 'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth', 'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1', 'AcceptedCmp2', 'Complain', 'Response']


/usr/local/lib/python3.12/dist-packages/sklearn/compose/_column_transformer.py:1667: FutureWarning: 
The format of the columns of the 'remainder' transformer in ColumnTransformer.transformers_ will change in version 1.7 to match the format of the other transformers.
At the moment the remainder columns are stored as indices (of type int). With the same ColumnTransformer configuration, in the future they will be stored as column names (of type str).
To use the new behavior now and suppress this warning, use ColumnTransformer(force_int_remainder_cols=False).

  warnings.warn(


Pipeline(steps=[('preprocess',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('cat',
                                                  OneHotEncoder(drop='first',
                                                                handle_unknown='ignore',
                                                                sparse_output=False),
                                                  ['Education',
                                                   'Marital_Status'])])),
                ('model',
                 RandomForestRegressor(max_depth=10, min_samples_leaf=5,
                                       min_samples_split=10,
                                       random_state=20))])

(iii). Use the model to **predict** the spending amount on wines by a customer with the following features.

| Feature |  Value  |
|---------|---------|
| Age     | 48      |
|Education|Graduation|
|Marital_Status| Married|
|Income|80,000|
|Kidhome|1|
|Teenhome|1|
|Dt-Customer|2016-10-10|
|Recency|43|     
|NumDealsPurchases|2|
|   NumWebPurchases|1|
|NumCatalogPurchases|0|
|NumStorePurchases|15|
|NumWebVisitsMonth|5|
|AcceptedCmp1,2,3,4,5| 0 |
|Complain|0|




In [25]:
# Create the new customer data point as specified
new_customer = pd.DataFrame([{
    'Year_Birth': 1977, #2025-48
    'Education': 'Graduation',
    'Marital_Status': 'Married',
    'Income': 80000.0,
    'Kidhome': 1,
    'Teenhome': 1,
    'Dt_Customer': (datetime.today() - pd.to_datetime('2016-10-10')).days,
    'Recency': 43,
    'NumDealsPurchases': 2,
    'NumWebPurchases': 1,
    'NumCatalogPurchases': 0,
    'NumStorePurchases': 15,
    'NumWebVisitsMonth': 5,
    'AcceptedCmp1': 0,
    'AcceptedCmp2': 0,
    'AcceptedCmp3': 0,
    'AcceptedCmp4': 0,
    'AcceptedCmp5': 0,
    'Response': 0,
    'Complain': 0
}])



# Make prediction
wine_prediction = pipe.predict(new_customer)
print(f"\nPredicted wine spending for the new customer: ${wine_prediction[0]:.2f}")


Predicted wine spending for the new customer: $389.85


(iv). **Consider** **two** measures to evaluate the model's performance on the test dataset.

Based on you computational results, how would you describe the model's performance?

In [26]:
# Make predictions on test set
y_pred_test = pipe.predict(X_test)

# Calculate evaluation metrics
# For regression, we'll use R² score and (R)MSE
r2 = r2_score(y_test, y_pred_test)
rmse = np.sqrt(mean_squared_error(y_test, y_pred_test))
mae = mean_absolute_error(y_test, y_pred_test)


print(f"R² Score: {r2:.4f}")
print(f"MAE: ${mae:.2f}")
print(f"RMSE: ${rmse:.2f}")

R² Score: 0.8140
MAE: $86.69
RMSE: $150.61


The R² score indicates that our Random Forest model explains a certain proportion of
the variance in wine spending. Based on the value obtained for R² of 0.8140, we can conclude there is a strong fit to the data. Model explains 81,40% of the variance


The RMSE tells us that on average, our predictions deviate from the actual wine
spending by 150 dollars. Given that wine spending in our dataset likely ranges from $0 to well over $1000, this error seems to be relatively good error margins.

The MAE provides another perspective, it tells us the typical absolute error we can
expect. This metric is less sensitive to outliers than RMSE, so it gives us a sense
of the "typical" prediction error. The 87 dollars our model gives us is quite reasonable.

Overall, looking at the combination of metrics, the model shows solid performance. The R² suggests good predictive power, though there's still
a slight bit of room for improvement unexplained variance. This could be due to inherent randomness in
customer behavior or unmeasured factors that influence wine purchases. For a retail
context where we're trying to understand customer spending patterns, this level of
accuracy provides an excellent foundation for making informed business decisions about
inventory management and customer targeting.

(v). **Perform** a 6-fold cross validation with a performance score of your choice.

**Note**: You may need to research on how to specify the performance score for regression models.

In [27]:
# We'll use R² as our scoring metric for cross-validation
# R² is intuitive for regression it tells us the proportion of variance explained
kf = KFold(n_splits=6, shuffle=True, random_state=20) # To ensure shuffling when CV

# Perform cross-validation
# We use the entire dataset (X, y) for cross-validation, not just training data
cv_scores = cross_val_score(
    pipe, X, y,
    cv=kf,
    scoring='r2',  # For regression, 'r2', 'neg_mean_squared_error', etc.
    n_jobs=-1  # Use all CPU cores for faster computation
)

print(f"R² scores for each fold: {cv_scores}")
print(f"Average R² score: {cv_scores.mean():.4f}")
print(f"Standard deviation: {cv_scores.std():.4f}")

R² scores for each fold: [0.7730935  0.79860186 0.75326097 0.82810769 0.80295795 0.78651103]
Average R² score: 0.7904
Standard deviation: 0.0236


(vi). **Perform** hyperparameter tuning using `GridSearchCV` for the following hyperparameters:

* Number of trees: 50, 100
* Maximum depth of any tree: 5, 10, 15
* Minimum number of data points required to split a node: 3, 6
* Minimum number of data points in any leaf node: 2,4,8


Based on your computational result, **show**:
* the best hyperparameter comination
* the corresponding performance score

In addition, **retrieve** the best model (the one corresponding to the best performance score).

In [28]:
# Define the parameter grid for GridSearchCV
# The keys must be in format 'model__parameter_name' since we're using a pipeline
param_grid = {
    'model__n_estimators': [50, 100],           # Number of trees
    'model__max_depth': [5, 10, 15],            # Maximum depth
    'model__min_samples_split': [3, 6],         # Min samples to split
    'model__min_samples_leaf': [2, 4, 8]        # Min samples per leaf
}

print(f"Parameter grid: {param_grid}")
print(f"Total combinations to test: {len(param_grid['model__n_estimators']) * len(param_grid['model__max_depth']) * len(param_grid['model__min_samples_split']) * len(param_grid['model__min_samples_leaf'])}")

# Initialize GridSearchCV
# We'll use 6-fold CV within the grid search for consistency
grid_search = GridSearchCV(
    estimator=pipe,
    param_grid=param_grid,
    cv=KFold(n_splits=6, shuffle=True, random_state=20),
    scoring='r2',  # Optimize for R² score
    n_jobs=-1,     # Use all available cores
    verbose=1      # Show progress
)

# Fit grid search on training data
grid_search.fit(X_train, y_train)


print(f"Best hyperparameters: {grid_search.best_params_}")
print(f"Best cross-validation R² score: {grid_search.best_score_:.4f}")

# Retrieve the best model
best_model = grid_search.best_estimator_

Parameter grid: {'model__n_estimators': [50, 100], 'model__max_depth': [5, 10, 15], 'model__min_samples_split': [3, 6], 'model__min_samples_leaf': [2, 4, 8]}
Total combinations to test: 36
Fitting 6 folds for each of 36 candidates, totalling 216 fits
Best hyperparameters: {'model__max_depth': 15, 'model__min_samples_leaf': 2, 'model__min_samples_split': 3, 'model__n_estimators': 100}
Best cross-validation R² score: 0.7865


In [29]:
# Evaluate best model on test set
y_pred_best = best_model.predict(X_test)
r2_best = r2_score(y_test, y_pred_best)
rmse_best = np.sqrt(mean_squared_error(y_test, y_pred_best))

print(f"\nBest model performance on test set:")
print(f"R² Score: {r2_best:.4f}")
print(f"RMSE: ${rmse_best:.2f}")

# View all results from grid search
results_df = pd.DataFrame(grid_search.cv_results_)
print(f"\nGrid search explored {len(results_df)} combinations")
print(results_df[['params', 'mean_test_score', 'std_test_score', 'rank_test_score']].sort_values('rank_test_score').head(10))


Best model performance on test set:
R² Score: 0.8296
RMSE: $144.17

Grid search explored 36 combinations
                                               params  mean_test_score  \
25  {'model__max_depth': 15, 'model__min_samples_l...         0.786512   
27  {'model__max_depth': 15, 'model__min_samples_l...         0.785562   
24  {'model__max_depth': 15, 'model__min_samples_l...         0.784437   
26  {'model__max_depth': 15, 'model__min_samples_l...         0.784245   
13  {'model__max_depth': 10, 'model__min_samples_l...         0.783805   
15  {'model__max_depth': 10, 'model__min_samples_l...         0.782857   
12  {'model__max_depth': 10, 'model__min_samples_l...         0.782323   
14  {'model__max_depth': 10, 'model__min_samples_l...         0.780958   
28  {'model__max_depth': 15, 'model__min_samples_l...         0.780877   
30  {'model__max_depth': 15, 'model__min_samples_l...         0.780877   

    std_test_score  rank_test_score  
25        0.024183                1  
27 

# Question 2 (24 points)

In this question, we will compare the performance of the best model found through `GridSearchCV` in Question 1 with the performance of the linear regression model.

(i). **Construct** a linear regression model using all relevant features and fit it to the training data.

Further, **evaluate** the model's performance on the test data and compare it with the best random forest model found in Question 1, with respect to the two performance considered in Question 1.

**Note**: You may use the same training and testing datasets as in Question 1.

In [33]:
# Create linear regression model
lr_model = LinearRegression()

# Create a pipeline with the same preprocessing
pipe_lr = Pipeline(steps=[
    ("preprocess", ct),
    ("model", lr_model)
])

# Fit to training data
pipe_lr.fit(X_train, y_train)

# Make predictions on test set
y_pred_lr = pipe_lr.predict(X_test)

# Calculate performance metrics for linear regression
r2_lr = r2_score(y_test, y_pred_lr)
rmse_lr = np.sqrt(mean_squared_error(y_test, y_pred_lr))
mae_lr = mean_absolute_error(y_test, y_pred_lr)

# Compare the two models
print("           Best LR left, Best RF right")
print(f"{'R² Score':<20} {r2_lr:<25.4f} {r2_best:<25.4f}")
print(f"{'RMSE':<20} ${rmse_lr:<24.2f} ${rmse_best:<24.2f}")
print(f"{'MAE':<20} ${mae_lr:<24.2f} ${mean_absolute_error(y_test, y_pred_best):<24.2f}")


           Best LR left, Best RF right
R² Score             0.7239                    0.8296                   
RMSE                 $183.51                   $144.17                  
MAE                  $119.03                   $80.97                   


(ii). Let's further compare the distribution of prediction errors by the two models in the following steps.

**Step 1**. For both the linear regression and the (best) random forest model, compute the absolute residual residual for each record in the test dataset.
  * Note that the absolute residual is distance between the predicted value and actual value, i.e., $|y_{pred}-y_{test}|$.

So we end up with two sets of absolute residuals (one by the linear regression model and the other by the random forest model).

**Step 2**. For each pair of absolute residuals for the same test data point, we can define a point in a scatterplot. Genereate such a scatterplot using `plotly.express` (LR residuals vs. RF residuals).  

**Step 3**. Add a 45 degree reference line to the plot. This can be done using the following codes. (You may need to change `min_val` and `max_val` for better visualization).

```
min_val = 0
max_val = 10
fig.add_shape(
    type="line",
    x0=min_val, y0=min_val,
    x1=max_val, y1=max_val,
    line=dict(color="red", width=2, dash="dash"),
)
```

**Implement the above steps**.

Note that the above steps essentially creates a [Q-Q plot](https://en.wikipedia.org/wiki/Q%E2%80%93Q_plot). How would you interpret the plot (for comparing the two predictive models).

In [31]:
# Step 1: Compute absolute residuals for both models
abs_residuals_lr = np.abs(y_test - y_pred_lr)
abs_residuals_rf = np.abs(y_test - y_pred_best)

# Create a DataFrame for plotting
comparison_df = pd.DataFrame({
    'Linear_Regression_Residual': abs_residuals_lr,
    'Random_Forest_Residual': abs_residuals_rf
})

# Step 2: Create scatter plot
fig = px.scatter(
    comparison_df,
    x='Linear_Regression_Residual',
    y='Random_Forest_Residual',
    title='Q-Q Plot: Comparison of Absolute Residuals (LR vs RF)',
    labels={
        'Linear_Regression_Residual': 'Linear Regression Absolute Residual ($)',
        'Random_Forest_Residual': 'Random Forest Absolute Residual ($)'
    },
    opacity=0.6
)

# Step 3: Add 45-degree reference line
min_val = 0
max_val = max(comparison_df['Linear_Regression_Residual'].max(),
              comparison_df['Random_Forest_Residual'].max())

fig.add_shape(
    type="line",
    x0=min_val, y0=min_val,
    x1=max_val, y1=max_val,
    line=dict(color="red", width=2, dash="dash"),
)

fig.show()

# Question 3. (24 points)

In this question, we will consider a classification problem on customers' spendings on meat products.


(i). For the column that represents customers' spendings on meat products, **calculate** the 33.33% and 66.67% percentiles. (**Hint**: You may use the function `df['MntMeatProducts'].quantile([1/3,2/3])`.)

Based on the two percentiles, **label** each row in the dataset
* If a customer's spending is below the 33.33% percentile, label their spending as "low";

* If a customer's spending is above the 66.67% percentile, label their spending as "high";


* If a customer's spending is between the two percentiles, label their spending as "medium".

(ii). **Build a K-nearest-neighbors model** to predict the label of a customer's spending on meat products. The model should be part of a machine learning pipeline that preprocesses the data.





(iii). Evaluate the performace of your model in part (ii) on the test data by **generating the classification report**.

Further, **interpret** each number in the classification report based on the current context.

# Describe how you used Gen. AI. in this assignment (2 points)

GenAI, in the form of Gemini, Perplexity and summaries generated durig Google searches, were used to troubleshoot errors and fix buggy code effectively, allowing for a quicker resolution of the issues. It also helped me with formatting some of the outputs in a way that is clearer to the user. eg: model comparison metrics

**Note**: The remaining 5 points will be assigned to readability of the work.